### ISSUE TITLE
send password-protected message through FES and update email format

### ISSUE THREAD

**Issue by DenBond7** [FlowCrypt/flowcrypt-android#1622]

**Comment by tomholub** [comment]
same as https://github.com/FlowCrypt/flowcrypt-ios/issues/1254#issue-1088458991

**Comment by DenBond7** [comment]
@tomholub 
Need to discuss a logic that will be integrated into the current implementation.

What we have now:

1. A user can create different types of a message - reply/reply all/forward with/without attachments(forwarded or attached)
2. Then he can click on the `Send` button and add a new message to the sending queue
3. Then in [PrepareOutgoingMessagesJobIntentService](https://github.com/FlowCrypt/flowcrypt-android/blob/72cab1e979bcfcb9ec67ed356e609617f7c869e0/FlowCrypt/src/main/java/com/flowcrypt/email/service/PrepareOutgoingMessagesJobIntentService.kt#L58) we prepare a new outgoing message: we create a raw MIME message with encrypted or not content(also we encrypt not forwarded attachments here)
4. If we have forwarded attachments we download them firstly ([ForwardedAttachmentsDownloaderWorker](https://github.com/FlowCrypt/flowcrypt-android/blob/72cab1e979bcfcb9ec67ed356e609617f7c869e0/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/ForwardedAttachmentsDownloaderWorker.kt#L57))
5. Then [MessagesSenderWorker](https://github.com/FlowCrypt/flowcrypt-android/blob/72cab1e979bcfcb9ec67ed356e609617f7c869e0/FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/MessagesSenderWorker.kt#L95) sends outgoing messages.

Need to think at what stage we should send a message via FES. For example, it can be `PrepareOutgoingMessagesJobIntentService`. But if we have the forwarded attachments we will have duplicate logic. In that case, `MessagesSenderWorker` can be the best place for `sending a message via FES`.

**Comment by tomholub** [comment]
This is up to you. It may turn out that the current structure is not the best fitting - that sometimes happens when new functionality needs to be added.

While uploading to FES, you'll have to make sure to continuously render displayed loader percentage. Can that be done from a service? And also from a worker?

It seems to me that, if you wanted to fit it into your structure, then `PrepareOutgoingMessagesJobIntentService` would be the right place to upload to FES because only after you get the URL from FES can you "prepare" the actual message that goes through Gmail. So the result would be truly a prepared message, which would be sent through `MessageSenderWorker` as is.

But how that plays together with forwarded attachments, I'm not sure. If you only download them in step 4, then you'd need to upload to FES in step 5 but that goes against your current design.

Can you switch step 4 (attachments) and step 3 (prepare)?

**Comment by DenBond7** [comment]
> Can you switch step 4 (attachments) and step 3 (prepare)?

of course. I added `ForwardedAttachmentsDownloaderWorker` just to reduce sending time for other outgoing messages.

How it works now:

1. Let's imagine a user has a slow connection(2G).
2. He is going to forward a message with a big attachment(around 20Mb)
3. He creates a forwarded message and clicks on the `Send` button
4. After that `PrepareOutgoingMessagesJobIntentService` creates a `MIME` message without attachments and mark a message as `MessageState.NEW_FORWARDED` and stops itself
5. After that `ForwardedAttachmentsDownloaderWorker` starts and begins to load forwarded attachments(it can take minutes with a slow connection). And marks the message as `MessageState.QUEUED` (ready for sending)
6. Before `5.` or during `5. `the user can create a new small message that will be created with `PrepareOutgoingMessagesJobIntentService` and marked as `MessageState.QUEUED`
7. `MessageSenderWorker` sends outgoing messages.

All of that means that the user will not wait while a big forwarded attachment will be downloaded. All messages without forwarded attachments will be sent almost immediately.

I'm still thinking about what to do and where.

**Comment by DenBond7** [comment]
@tomholub 

>While uploading to FES, you'll have to make sure to continuously render displayed loader percentage. Can that be done from a service? And also from a worker?

In that case, I can do all things with FES interactively. I can show a dialog with progress. When a message will be uploaded to the FES, I will have `URL` and will be able to add a new message to the queue.

**procs**

- we will show progress
- we don't modify the existing sending process (almost)

**cons**

- a user should wait while the current process will be completed to be able to make any actions with the app.

**Comment by IvanPizhenko** [comment]
Isn't it better to have some "background" notification with progress? Like, for example, OneDrive or Dropbox? The have user notification with progress, while this doesn't block other use of application.

**Comment by tomholub** [comment]
> In that case, I can do all things with FES interactively. I can show a dialog with progress. When a message will be uploaded to the FES, I will have URL and will be able to add a new message to the queue.

I see - the actual sending, though Gmail later, happens in background queue on Android. So there would be a foreground loader for FES upload, and then background sending through Gmail, so you don't need to do the 50% progress.

> Isn't it better to have some "background" notification with progress? Like, for example, OneDrive or Dropbox? The have user notification with progress, while this doesn't block other use of application.

Agree it would be better. It could be left for later as a separate issue, if it's more work to do. 

Anyway, on iOS, we're also doing a foreground wait for now. Not ideal, but it saves work (for now).

**Comment by DenBond7** [comment]
I'm asking just in case. Why it is so important to show progress(with percentage) for sending via `FES`? I mean we don't think about that for a regular sending via `IMAP/GMail API`. A user sends a message to a queue of outgoing messages. After that, we show the following notification

![image](https://user-images.githubusercontent.com/2863246/147653144-6010ea10-40b4-443c-bd7f-b8137fe5f630.png)

It means the app is sending outgoing messages. This notification can cover this scenario too.

We don't show progress for every separate message in notifications. A user can check the current state of messages in the Outbox folder.

![image](https://user-images.githubusercontent.com/2863246/147653506-a4a187d1-7aa5-49f6-84a7-229a64a2ead2.png)

If we use the `queue` strategy I'm not sure that it is so important to show progress with percentage. It's cool, but maybe as improvements for later.

The `queue` strategy means we send a message to a queue and forget about it. Just the app should notify us if an error has occurred.

**Comment by IvanPizhenko** [comment]
@DenBond7 is it significantly harder to have numeric progress in the notification? such as "Sending emails (1/100) ..."?

**Comment by DenBond7** [comment]
> is it significantly harder to have numeric progress in the notification? such as "Sending emails (1/100) ..."?

nope, we can do that without big changes.

**Comment by tomholub** [comment]
> The queue strategy means we send a message to a queue and forget about it. Just the app should notify us if an error has occurred.

The task definition was assuming that user waits for the sending (which is the case on iOS). Since on Android we have a background queue, it's different, so the percentages are less important.

**Comment by DenBond7** [comment]
That's good news! I already have a simple testable version that uses the same behavior as `ForwardedAttachmentsDownloaderWorker`

**Comment by tomholub** [comment]
Nevertheless, let's file Ivan's suggestion as a separate issue for `later` milestone.

**Comment by DenBond7** [comment]
> Nevertheless, let's file Ivan's suggestion as a separate issue for `later` milestone.

Agree. This issue will cover only basic functionally

**Comment by DenBond7** [comment]
test the following cases
- [ ] only message
- [ ] message + att
- [ ] message + att + forwarded att
